File: aqrar_ext/public/js/sales_order_payment.js
// aqrar_ext/public/js/sales_order_payment.js
//
// "Receive Payment" on a submitted Sales Order — the order-side twin of the
// invoice popup in sales_invoice_pos_total_popup.js. Ported from Steel Force's
// public/js/sales_order_payment.js and kept deliberately consistent with the
// invoice dialog already running here: one amount row per allowed mode of
// payment, cheque number/date asked for only when a cheque mode carries an
// amount, and the whole collection sent in a single call.
//
// The server (aqrar_ext/api/sales_order_payment.py) is the authority on what may
// be collected; everything here is convenience and early feedback.

frappe.ui.form.on("Sales Order", {
    refresh(frm) {
        if (frm.doc.docstatus !== 1) return;
        // The server refuses both, so offering the dialog on a Closed / On Hold
        // order would only collect keystrokes it is going to throw away.
        if (frm.doc.status === "Closed" || frm.doc.status === "On Hold") return;

        frm.add_custom_button(__("Receive Payment"), function () {
            aqrar_show_so_payment_popup(frm);
        }).addClass("btn-primary");
    },
});

function aqrar_so_precision(state) {
    const p = cint(state && state.precision);
    return p > 0 ? p : 2;
}

function aqrar_show_so_payment_popup(frm) {
    if (frappe.flags.aqrar_so_popup_showing) return;
    frappe.flags.aqrar_so_popup_showing = true;

    const release = () => { frappe.flags.aqrar_so_popup_showing = false; };

    frappe.call({
        method: "aqrar_ext.api.sales_order_payment.get_sales_order_payment_state",
        args: { sales_order: frm.doc.name },
        freeze: true,
        freeze_message: __("Loading payment modes..."),
        callback: function (r) {
            const state = r && r.message;
            if (!state) return release();

            if (flt(state.balance) <= 0) {
                release();
                frappe.msgprint({
                    title: __("Nothing to Collect"),
                    message: __("Order {0} is already fully paid in advance.", [frm.doc.name]),
                    indicator: "blue",
                });
                return;
            }

            const modes = state.modes || [];
            const pdc_modes = state.pdc_modes || [];
            if (!modes.length && !pdc_modes.length) {
                release();
                frappe.msgprint({
                    title: __("No Payment Modes"),
                    message: __(
                        "No mode of payment is configured for your branch on company {0}.",
                        [state.company]
                    ),
                    indicator: "red",
                });
                return;
            }
            aqrar_render_so_dialog(frm, state, release);
        },
        error: release,
    });
}

function aqrar_render_so_dialog(frm, state, release) {
    const precision = aqrar_so_precision(state);
    const modes = state.modes || [];
    const pdc_modes = state.pdc_modes || [];
    const all_modes = modes.concat(pdc_modes.filter((m) => modes.indexOf(m) === -1));
    const is_cheque_mode = (m) => pdc_modes.indexOf(m) !== -1;

    const fields = [
        {
            fieldtype: "HTML",
            fieldname: "summary",
            options: `
                <div style="padding:8px 0 12px 0;line-height:1.7">
                    <div><b>${__("Order Total")}:</b>
                        ${format_currency(state.grand_total, state.currency)}</div>
                    <div><b>${__("Already Advanced")}:</b>
                        ${format_currency(state.advance_paid, state.currency)}</div>
                    <div style="font-size:1.05em"><b>${__("Balance to Collect")}:</b>
                        <span style="color:var(--text-color)">
                        ${format_currency(state.balance, state.currency)}</span></div>
                </div>`,
        },
        { fieldtype: "Section Break", label: __("Amounts") },
    ];

    all_modes.forEach((mode, i) => {
        fields.push({
            fieldtype: "Currency",
            fieldname: `amt_${i}`,
            label: mode + (is_cheque_mode(mode) ? ` (${__("Cheque")})` : ""),
            default: 0,
            precision: precision,
            onchange: function () { recompute(); },
        });
        if (i % 2 === 0) fields.push({ fieldtype: "Column Break" });
    });

    fields.push(
        { fieldtype: "Section Break" },
        {
            // Drives depends_on for the cheque fields. A real (hidden) field is
            // required: names beginning "__" are stripped from the dialog doc, so
            // set_value on one silently does nothing and the fields never appear.
            fieldtype: "Check",
            fieldname: "has_cheque",
            label: "has_cheque",
            hidden: 1,
            default: 0,
        },
        {
            fieldtype: "Date",
            fieldname: "posting_date",
            label: __("Posting Date"),
            default: frappe.datetime.get_today(),
        },
        { fieldtype: "Column Break" },
        {
            fieldtype: "Data",
            fieldname: "cheque_no",
            label: __("Cheque No"),
            depends_on: "eval:doc.has_cheque",
        },
        {
            fieldtype: "Date",
            fieldname: "cheque_date",
            label: __("Cheque Date"),
            depends_on: "eval:doc.has_cheque",
        },
        { fieldtype: "Section Break" },
        { fieldtype: "HTML", fieldname: "totals" }
    );

    const dialog = new frappe.ui.Dialog({
        title: __("Receive Payment — {0}", [frm.doc.name]),
        size: "large",
        fields: fields,
        primary_action_label: __("Receive Payment"),
        primary_action: function (vals) { submit(vals); },
    });

    dialog.onhide = () => release();

    function collected() {
        const rows = [];
        all_modes.forEach((mode, i) => {
            const amt = flt(dialog.get_value(`amt_${i}`));
            if (amt > 0) rows.push({ mode_of_payment: mode, amount: amt });
        });
        return rows;
    }

    function recompute() {
        const rows = collected();
        const total = rows.reduce((s, r) => s + flt(r.amount), 0);
        const remaining = flt(state.balance) - total;
        const has_cheque = rows.some((r) => is_cheque_mode(r.mode_of_payment));

        // depends_on reads the dialog doc, so the flag has to live on it
        dialog.set_value("has_cheque", has_cheque ? 1 : 0);
        dialog.refresh();

        const over = remaining < -0.0001;
        dialog.fields_dict.totals.$wrapper.html(`
            <div style="padding:6px 0;line-height:1.8">
                <div><b>${__("Entered")}:</b> ${format_currency(total, state.currency)}</div>
                <div style="color:${over ? "var(--red-500)" : "var(--text-muted)"}">
                    <b>${over ? __("Over by") : __("Remaining")}:</b>
                    ${format_currency(Math.abs(remaining), state.currency)}
                </div>
                ${over
                    ? `<div style="color:var(--red-500)">${__(
                          "The total is more than the order balance."
                      )}</div>`
                    : ""}
            </div>`);
    }

    function submit(vals) {
        const rows = collected();
        if (!rows.length) {
            frappe.msgprint(__("Enter an amount against at least one mode of payment."));
            return;
        }
        const total = rows.reduce((s, r) => s + flt(r.amount), 0);
        if (total - flt(state.balance) > 0.0001) {
            frappe.msgprint(
                __("Total {0} is more than the balance of {1}.", [
                    format_currency(total, state.currency),
                    format_currency(state.balance, state.currency),
                ])
            );
            return;
        }
        const needs_cheque = rows.some((r) => is_cheque_mode(r.mode_of_payment));
        if (needs_cheque && !vals.cheque_no) {
            frappe.msgprint(__("Enter the cheque number."));
            return;
        }

        dialog.disable_primary_action();
        frappe.call({
            method: "aqrar_ext.api.sales_order_payment.create_payments_for_sales_order",
            args: {
                sales_order: frm.doc.name,
                payments: JSON.stringify(rows),
                cheque_no: vals.cheque_no || null,
                cheque_date: vals.cheque_date || null,
                posting_date: vals.posting_date || null,
            },
            freeze: true,
            freeze_message: __("Creating Payment Entries..."),
            callback: function (r) {
                const created = (r && r.message) || [];
                dialog.hide();
                frm.reload_doc();
                frappe.show_alert(
                    {
                        message: __("{0} Payment Entry(s) created: {1}", [
                            created.length,
                            created.join(", "),
                        ]),
                        indicator: "green",
                    },
                    7
                );
            },
            error: function () { dialog.enable_primary_action(); },
        });
    }

    dialog.show();
    recompute();
}

File: aqrar_ext/api/sales_order_payment.py
# aqrar_ext/api/sales_order_payment.py
"""Collect money against a Sales Order, the same way the Sales Invoice popup does.

The order-side twin of ``aqrar_ext.api.sales_invoice``: the same branch allow-list
of modes of payment, the same one-Payment-Entry-per-mode result, the same cheque
handling. Ported from Steel Force's ``sf_trading/api/sales_order_payment.py``.

An order is not an invoice, so two things differ by design:

* **No Loyalty / write-off.** A write-off closes an outstanding invoice by booking
  the shortfall to the company's Write Off Account; an order has no receivable to
  close. Steel Force offers it; the Aqrar invoice port already dropped loyalty, so
  offering it here would be the only place in this app that has it.

* **"Amount to Pay" is the order's own balance** -- ``grand_total`` less
  ``advance_paid`` -- not an invoice's ``outstanding_amount``. ERPNext keeps
  ``advance_paid`` in the party account's currency and refreshes it from the
  Advance Payment Ledger Entry table on every payment submit, which is why the
  balance is re-read from the order between each mode rather than computed once.

Everything the resulting Payment Entry *is* comes from ERPNext's own
``get_payment_entry``, so the entries are indistinguishable from desk-made ones.

THREE ADAPTATIONS from the Steel Force original
-----------------------------------------------
1. **Branch.** Steel Force reads ``Sales Order.branch``. This site has no branch
   field on Sales Order (verified against the bench), so the branch is resolved
   from the user's own Branch Configuration via
   :func:`aqrar_ext.api.sales_invoice.resolve_branch` -- exactly what the invoice
   popup already does here.

2. **Payment Terms (a bug fixed, not merely ported).** When the order's Payment
   Terms Template has ``allocate_payment_based_on_payment_terms``, ERPNext's
   ``get_payment_entry`` appends ONE references row per payment term. The Steel
   Force version writes ``references[0]`` only and leaves rows 1..N at their full
   allocated amount, so a part payment allocates the whole order. Aqrar assigns a
   payment_terms_template from the Customer (see
   ``public/js/sales_invoice_payment_terms.js``), so this is reachable here.
   :func:`_spread_allocation` distributes the collected amount across the rows in
   order and drops the rows left at zero.

3. **No ZATCA prepayment invoice is raised.** ``ksa_compliance`` DOES ship a
   complete advance/prepayment path -- it registers Payment Entry ``validate`` /
   ``on_submit`` / ``before_cancel`` doc_events and emits a PREPAYMENT invoice
   when ``custom_prepayment_invoice`` is ticked on the Payment Entry. This module
   deliberately does NOT tick it, so an advance collected here posts the money and
   raises no e-invoice. Two reasons, both of which the business must decide before
   that changes:
     * whether a customer deposit against an order requires an advance tax invoice
       is a tax question, not an engineering one; and
     * ``ksa_compliance`` hard-blocks cancellation of a prepayment Payment Entry
       once Phase 2 is live, so a mis-keyed deposit becomes permanently
       uncancellable.
   Opting in means setting ``custom_prepayment_invoice``,
   ``custom_prepayment_invoice_description``, ``custom_posting_time`` and
   ``sales_taxes_and_charges_template`` on the Payment Entry built below.
"""

import json

import frappe
from frappe import _
from frappe.utils import cint, flt, nowdate

from aqrar_ext.api.sales_invoice import (
	branch_has_pdc_modes,
	get_payment_modes_with_account,
	resolve_branch,
)


def _currency_precision(currency):
	"""Decimals the currency is actually kept in (SAR is two, BHD three)."""
	number_format = frappe.db.get_value("Currency", currency, "number_format") or "#,###.##"
	return len(number_format.split(".")[-1]) if "." in number_format else 0


def _order_balance(so):
	"""What is still uncollected on the order, clamped at zero.

	``advance_paid`` is denominated in the party account's currency while
	``grand_total`` is in the order's own. They only differ on a foreign-currency
	order; this site sells in company currency, so they are compared directly.
	"""
	total = flt(so.get("rounded_total") or so.get("grand_total"))
	return max(0.0, flt(total - flt(so.get("advance_paid")), _currency_precision(so.currency)))


@frappe.whitelist()
def get_sales_order_payment_state(sales_order):
	"""Everything the Receive Payment dialog needs, in one round trip.

	Modes come back split the way the invoice popup splits them -- the branch's
	cheque (``for_pdc``) modes apart from the rest -- so the dialog asks for a
	cheque number only when a cheque amount is actually entered.
	"""
	frappe.has_permission("Sales Order", "read", doc=sales_order, throw=True)

	so = frappe.get_doc("Sales Order", sales_order)
	branch = resolve_branch(None, frappe.session.user)

	return {
		"sales_order": so.name,
		"company": so.company,
		"currency": so.currency,
		"precision": _currency_precision(so.currency),
		"grand_total": flt(so.get("rounded_total") or so.get("grand_total")),
		"advance_paid": flt(so.get("advance_paid")),
		"balance": _order_balance(so),
		"per_billed": flt(so.get("per_billed")),
		"status": so.get("status"),
		"modes": get_payment_modes_with_account(so.company, is_return=0, is_pdc=0, branch=branch),
		"pdc_modes": (
			get_payment_modes_with_account(so.company, is_return=0, is_pdc=1, branch=branch)
			if branch_has_pdc_modes(branch)
			else []
		),
	}


def _cheque_modes(company):
	"""The branch's cheque (``for_pdc``) modes, so only those carry the cheque date."""
	branch = resolve_branch(None, frappe.session.user)
	if not branch or not cint(branch_has_pdc_modes(branch)):
		return set()
	return set(get_payment_modes_with_account(company, is_return=0, is_pdc=1, branch=branch))


def _spread_allocation(pe, allocated, precision):
	"""Distribute ``allocated`` across the reference rows, in order.

	``get_payment_entry`` appends one references row per payment term when the
	order's Payment Terms Template allocates by term. Writing only row 0 -- what
	the Steel Force original does -- leaves the remaining rows at their full
	allocated amount, so a part payment allocates the entire order and ERPNext
	then refuses the entry (or accepts it and over-advances the order).

	Rows left at zero are dropped: ERPNext treats a zero-allocation reference row
	as an error on submit.
	"""
	remaining = flt(allocated, precision)
	for ref in pe.references:
		if remaining <= 0:
			ref.allocated_amount = 0
			continue
		take = min(flt(ref.allocated_amount), remaining)
		ref.allocated_amount = flt(take, precision)
		remaining = flt(remaining - take, precision)

	pe.references = [ref for ref in pe.references if flt(ref.allocated_amount) > 0]
	return pe.references


@frappe.whitelist()
def create_payments_for_sales_order(
	sales_order,
	payments,
	cheque_date=None,
	cheque_no=None,
	posting_date=None,
):
	"""One submitted Payment Entry per mode of payment, as an advance on the order.

	Args:
		sales_order: submitted Sales Order name.
		payments: JSON list (or list) of ``{"mode_of_payment": str, "amount": float}``.
		cheque_date / cheque_no: the cheque's own date and number. The posting date
			stays today's -- the cheque date rides on ``reference_date``.
		posting_date: overrides today, for a collection recorded after the fact.

	Returns:
		list[str]: created Payment Entry names.
	"""
	if not sales_order:
		frappe.throw(_("Sales Order is required"))

	frappe.has_permission("Sales Order", "read", doc=sales_order, throw=True)
	frappe.has_permission("Payment Entry", "create", throw=True)

	so = frappe.get_doc("Sales Order", sales_order)
	if so.docstatus != 1:
		frappe.throw(
			_("Sales Order {0} must be submitted before collecting payment.").format(so.name)
		)

	# A closed order has been abandoned and one on hold is disputed -- taking money
	# against either is a decision somebody must make deliberately, by reopening it.
	if so.status in ("Closed", "On Hold"):
		frappe.throw(
			_("Sales Order {0} is {1}. Reopen it before collecting a payment against it.").format(
				so.name, _(so.status)
			)
		)

	if isinstance(payments, str):
		try:
			payments = json.loads(payments)
		except Exception:
			frappe.throw(_("Invalid payments payload"))

	if not isinstance(payments, (list, tuple)) or not payments:
		frappe.throw(_("No payment rows were provided."))

	valid_rows = []
	for row in payments:
		mode_of_payment = (row or {}).get("mode_of_payment")
		amount = flt((row or {}).get("amount"))
		if not mode_of_payment or amount <= 0:
			continue
		valid_rows.append({"mode_of_payment": mode_of_payment, "amount": amount})

	if not valid_rows:
		frappe.throw(_("No valid payment rows found (non-zero amounts with mode of payment)."))

	precision = _currency_precision(so.currency)
	asked = flt(sum(row["amount"] for row in valid_rows), precision)
	balance_now = _order_balance(so)

	if asked - balance_now > 0.0001:
		frappe.throw(
			_("Total payment {0} is more than the balance of {1} on order {2}.").format(
				asked, balance_now, so.name
			)
		)

	from erpnext.accounts.doctype.payment_entry.payment_entry import get_payment_entry
	from erpnext.accounts.doctype.sales_invoice.sales_invoice import get_bank_cash_account

	cheque_modes = _cheque_modes(so.company)
	created = []
	for row in valid_rows:
		# Re-read the order: ERPNext refreshes `advance_paid` from the advance ledger
		# when a payment is submitted, so the balance the next mode may take is only
		# known now.
		so.reload()
		balance = _order_balance(so)
		amount = flt(row["amount"], precision)
		if amount <= 0 or balance <= 0:
			continue

		pe = get_payment_entry("Sales Order", so.name)
		pe.mode_of_payment = row["mode_of_payment"]

		bank_account = (
			get_bank_cash_account(row["mode_of_payment"], so.company) or {}
		).get("account")
		if not bank_account:
			frappe.throw(
				_("Set a default account for mode of payment {0} on company {1}.").format(
					row["mode_of_payment"], so.company
				)
			)
		pe.paid_to = bank_account
		account = frappe.get_cached_value(
			"Account", bank_account, ["account_currency", "account_type"], as_dict=True
		)
		if account:
			pe.paid_to_account_currency = account.account_currency
			pe.paid_to_account_type = account.account_type

		if not pe.references:
			frappe.throw(_("Order {0} has nothing left to collect against.").format(so.name))

		# Never allocate past the order's own balance. `outstanding_amount` on each row
		# is left exactly as ERPNext wrote it -- rewriting it trips "has already been
		# partly paid" on submit.
		fillable = flt(sum(flt(ref.allocated_amount) for ref in pe.references), precision)
		allocated = min(amount, balance, fillable)
		if allocated <= 0:
			continue

		_spread_allocation(pe, allocated, precision)
		pe.paid_amount = flt(allocated, precision)
		pe.received_amount = pe.paid_amount

		pe.posting_date = posting_date or nowdate()
		pe.reference_no = cheque_no or so.name
		if cheque_date and row["mode_of_payment"] in cheque_modes:
			pe.reference_date = cheque_date
		else:
			pe.reference_date = pe.posting_date

		pe.insert()

		# Same reasoning as the invoice popup: an entry the cashier never sees cannot
		# be walked through an approval chain, and the collection has already happened.
		pe.flags.ignore_workflow = True
		pe.flags.ignore_validate = True
		pe.submit()
		created.append(pe.name)

	return created
